astropy/modeling/functional_models.py

```diff
@@ -572,6 +572,104 @@ def fit_deriv(x, amplitude, x_0, fwhm):
         return [d_amplitude, d_x_0, d_fwhm]
 
 
+class Voigt1D(Fittable1DModel):
+    """
+    One dimensional model for the Voigt profile.
+
+    Parameters
+    ----------
+    x_0 : float
+        Position of the peak
+    amplitude_L : float
+        The Lorentzian amplitude
+    fwhm_L : float
+        The Lorentzian full width at half maximum
+    fwhm_G : float
+        The Gaussian full width at half maximum
+
+    See Also
+    --------
+    Gaussian1D, Lorentz1D
+
+    Notes
+    -----
+    Algorithm for the computation taken from
+    McLean, A. B., Mitchell, C. E. J. & Swanston, D. M. Implementation of an
+    efficient analytical approximation to the Voigt function for photoemission
+    lineshape analysis. Journal of Electron Spectroscopy and Related Phenomena
+    69, 125-132 (1994)
+
+    Examples
+    --------
+    .. plot::
+        :include-source:
+
+        import numpy as np
+        from astropy.modeling.models import Voigt1D
+        import matplotlib.pyplot as plt
+
+        plt.figure()
+        x = np.arange(0, 10, 0.1)
+        v1 = Voigt1D(x_0=5,amplitude_L=10,fwhm_L=0.5, fwhm_G=0.9)
+        plt.plot(x, v1(x))
+        plt.show()
+    """
+
+    x_0 = Parameter(default=0)
+    amplitude_L = Parameter(default=1)
+    fwhm_L = Parameter(default=2/np.pi)
+    fwhm_G = Parameter(default=np.log(2))
+
+    _abcd = np.array([
+        [-1.2150, -1.3509, -1.2150, -1.3509],  # A
+        [1.2359, 0.3786, -1.2359, -0.3786],    # B
+        [-0.3085, 0.5906, 0.3085, 0.5906],     # C
+        [0.0210, -1.1858, -0.0210, 1.1858]])   # D
+
+    @classmethod
+    def evaluate(cls, x, x_0, amplitude_L, fwhm_L, fwhm_G):
+
+        A, B, C, D = cls._abcd
+        sqrt_ln2 = np.sqrt(np.log(2))
+        X = (x - x_0) * 2 * sqrt_ln2 / fwhm_G
+        X = X[:, np.newaxis]
+        Y = fwhm_L * sqrt_ln2 / fwhm_G
+        Y = Y[:, np.newaxis]
+        V = np.sum((C * (Y - A) + D * (X - B))/(((Y - A) ** 2 + (X - B) ** 2)), axis=-1)
+
+        return (fwhm_L * amplitude_L * np.sqrt(np.pi) * sqrt_ln2 / fwhm_G) * V
+
+    @classmethod
+    def fit_deriv(cls, x, x_0, amplitude_L, fwhm_L, fwhm_G):
+
+        A, B, C, D = cls._abcd
+        sqrt_ln2 = np.sqrt(np.log(2))
+        a = [amplitude_L, x_0, fwhm_L, fwhm_G]
+        dVdx = 0
+        dVdy = 0
+        V = 0
+        dyda = [0, 0, 0, 0]
+
+        X = (x - a[1]) * 2 * sqrt_ln2 / (a[3])
+        X = X[:, np.newaxis]
```

Comment:
When I try fitting this I get an exception at this line.  I think there's a subtle difference between `evaluate` and `fit_deriv`, in that with `evaluate` the Model framework makes sure that all the values passed in when it's called are arrays.  Whereas with `fit_deriv` is called from the optimizer code it doesn't always make this guarantee.  That should probably be fixed on our end so that writing the `fit_derive` code can make the same assumptions.

In the meantime try `X = np.atleast_1d(X)[:, np.newaxis]` and similarly for `Y`.

I'm still not sure why it seems like it's not even calling it for you.